### ISSUE TITLE
Not receive Environment in nn.Module, Metrics

### ISSUE THREAD

**Issue by HiroakiMikami** [HiroakiMikami/mlprogram#143]
* Create utility for Environment
    * [x] `Apply` enable to store a value with multiple keys
    * [x] Move `Apply` and `Pick` to `mlprogram`
    * [x] `supervision` flag should be inherited via Apply
* [x] function objects inherits `nn.Module`  (transforms and metrics)
* [x] Rewrite functions
    * csg.TransformCanvas
    * ~~csg.AddTestCases~~
    * ~~linediff.AddTestCases~~
    * ~~linediff.UpdateInput~~
    * ~~nn.Accuracy~~
    * ~~nn.ActionSequenceReader~~
    * ~~nn.Loss~~
    * ~~nn.Predictor~~
    * ~~nn.RnnDecoder~~
    * ~~nn.Decoder~~
    * ~~nn.Encoder (pbe_with_repl)V~~
    * ~~nn.NlReader~~
    * ~~AddEmptyReference~~
    * ~~GroundTruthToActionSequence~~
    * ~~EncodeActionSequence~~
    * ~~AddPreviousActions~~
    * ~~AddActions~~
    * ~~AddPreviousActionRules~~
    * ~~AddActionSequenceAsTree~~
    * ~~AddQueryForTreeGenDecoder~~
    * ~~AddStateForRnnDecoder~~
    * ~~AddHistoryState~~
    * ~~NormalizeGroundTruth~~
    * ~~ExtractReference~~
    * ~~EncodeTokenQuery~~
    * ~~EncodeWordQuery~~
    * ~~EncodeCharacterQuery~~
    * metrics.Xxx
* [x] Remove `ExtractReference`
* [x] make `ToEpisode` nn.Module
